src/middleware/mod.rs

```diff
@@ -218,6 +230,32 @@ impl Value {
     pub fn raw(&self) -> RawValue {
         self.raw
     }
+    /// Determines Merkle existence proof for `key` in `self` (if applicable).
+    pub(crate) fn prove_existence<'a>(
+        &'a self,
+        key: &'a Value,
+    ) -> Result<Option<(&'a Value, MerkleProof)>> {
```

Comment:
Why not make this `Result<(&'a Value, MerkleProof)>` and return an error if `self` is not a container?
I think this change can simplify the code a little bit.